# Laboration 3 - Markovkedjor och Köteori

#### Eric Johansson, Can Kupeli, Samuel Greenberg

In [37]:
import math
import numpy as np 
import sympy as sy 
from matplotlib import pyplot as plt

## Uppgift 1

### a)

In [38]:
P = np.array([[0.80, 0.15, 0.05], [0.10, 0.75, 0.15], [0.25, 0.25, 0.50]])


Överföringsmatrisen från Figur 1

### b)

In [39]:
s = np.array([1, 0, 0])

tolerance = 1e-6
err = 1

iterations = 0

while err > tolerance:
    s_0 = s
    s = s_0 @ P
    err = np.max(np.abs(s - s_0))
    iterations += 1


print(s)
print(iterations)


[0.41666813 0.41666569 0.16666618]
30


### c)

Med toleransen $ 10^{-6} $ fick vi stationära lösningen [0.41666684 0.41666655 0.16666661]

### d)

Nej, oavsett startvektor kommer vi till samma sationära lösning. Det är således en egodisk process

### e)

Ja, Den snabbaste konvergensen får vi om vi använder den riktiga stationära lösningen som startvektor.

## Uppgift 2

In [27]:
from sympy.stats import DiscreteMarkovChain as dmc

T = sy.Matrix(
    [
        [4 / 5, 3 / 20, 1 / 20],
        [1 / 10, 3 / 4, 3 / 20],
        [1 / 4, 1 / 4, 1 / 2],
    ]
)
X = sy.stats.dmc("X", trans_probs=T)

print(X.stationary_distribution())


Matrix([[0.416666666666667, 0.416666666666667, 0.166666666666667]])


Ett exempel på hur man löser den med sympy. Från [symPy dokumentationen](https://docs.sympy.org/latest/modules/stats.html)

## Uppgift 3

In [28]:
F = np.cumsum(P, axis=1)

N = 10**6
counts = np.array([0, 0, 0])

real = [5 / 12, 5 / 12, 1 / 6]
err = 1

bull_index = 0
bear_index = 1
rec_index = 2

curr = bull_index

for _ in range(N):
    counts[curr] = counts[curr] + 1
    rnd = np.random.rand()
    if rnd < F[curr, bull_index]:
        curr = bull_index
    elif rnd < F[curr, bear_index]:
        curr = bear_index
    else:
        curr = rec_index

print(
    f"Med {N} upprepningar fick vi sannolikheterna {counts/np.sum(counts)}"
)


Med 1000000 upprepningar fick vi sannolikheterna [0.415997 0.417472 0.166531]


## Uppgift 4

### a)

In [29]:
Periodisk = [[0, 1, 0], [0, 0, 1], [1, 0, 0]]

En cyklisk kedja är en som konstant går runt och runt. Denna kommmer inte att komma till en stationär lösning med hjälp av metoden i uppgift 1

### b)

In [30]:
Reducibel = [[0.5, 0.5, 0], [0.5, 0.5, 0], [0, 0, 1]]


Denna matris är icke-ergodisk då startvektorn kommer ge olika resultat beroende på om man startar på de första två eller den sista

### c)

Ett exempel på en icke-minnesfri process vore om man inte kan åka tillbaka till stället man kom ifrån. Då måste man ha stället man kom ifrån sparat i minnet för att justera sannolikhetena

### d)

Fördelningen i (a) skulle kunna beräknas genom att räkna hur ofta vi är vid varje punkt och sedan dividera med antalet för att få någon slags medelvärde.

I (b) blir man tvingad att komplettera med en yttre loop med flera olika starttillstånd och beräkna ensemblevärdet.

Fördelningen i (c) kommer kräva stor ändring av koden och kan därför inte lätt använda sig av metoderna i uppgift 1-3

## Uppgift 5

### a)

The M/M/2/4 is a Kendall notation, the first two variables represent the arrival and departure. The M's stands for Markov process, a poisson arrival (negative exponetial service). The third variable represents the number of servers, in this case 2. The fourth and final variable represents the buffer, the maximum number of customers allowed in the system, in this case 4.

### b)

![](5b.JPG "Markov Chain")
Markov Chains state diagram with transition probabilities

### c)

In [31]:
lam = 0.3;
mu = 0.2;

P = np.array([
  [0.7, 0.3, 0.0, 0.0, 0.0],
  [0.2, 0.5, 0.3, 0.0, 0.0],
  [0.0, 0.4, 0.3, 0.3, 0.0],
  [0.0, 0.0, 0.4, 0.3, 0.3],
  [0.0, 0.0, 0.0, 0.4, 0.6]
])

### d)

In [32]:
T = sy.Matrix(P)
X = sy.stats.dmc("X", trans_probs=T)

real = np.array(X.stationary_distribution()).astype(np.float64)
print(real)

[[0.19601838 0.29402757 0.22052067 0.16539051 0.12404288]]


### e)

In [33]:
L = np.sum(np.array([0,1,2,3,4]).T * real);
print(L)

1.7274119448698317


### f)

In [34]:
LittlesLaw = L/lam
print(LittlesLaw)

5.758039816232773


Using Little's Law the average time a customer spends within the system is calculated to 5.76 minutes. 
This was done by dividing the average number of customers in the system by the customer arrival rate.